misc/multistream-select/src/protocol/dialer.rs

```diff
@@ -64,24 +67,22 @@ where
     R: AsyncRead + AsyncWrite,
     N: AsRef<[u8]>
 {
-    type SinkItem = DialerToListenerMessage<N>;
+    type SinkItem = Request<N>;
     type SinkError = MultistreamSelectError;
 
-    #[inline]
-    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        match self.inner.start_send(Message::Body(item))? {
-            AsyncSink::NotReady(Message::Body(item)) => Ok(AsyncSink::NotReady(item)),
-            AsyncSink::NotReady(Message::Header) => unreachable!(),
-            AsyncSink::Ready => Ok(AsyncSink::Ready)
+    fn start_send(&mut self, request: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        let mut msg = BytesMut::new();
+        request.encode(&mut msg)?;
+        match self.inner.start_send(msg.freeze())? {
+            AsyncSink::NotReady(_) => Ok(AsyncSink::NotReady(request)),
```

Comment:
For what it's worth, in new futures you have to wait for the `Sink` to be ready before calling `start_send`. Calling `start_send` is then guaranteed to never fail for readiness-related reasons.

Reply:
Just curious: with the current futures, do you see a better way for the general case of wanting to nest `Sink`s with different item types (due to being different layers of abstraction) while avoiding the re-encoding from the upper to the lower-level item type? I guess if the encoding would be expensive, an additional buffer of encoded messages could do? In any case, I obviously didn't think this an important issue here since the encoding is presumably  cheap and from my understanding typically at most done twice for 1 out of N > 1 messages, assuming the sinks can generally buffer quite a few messages to send before being `NotReady`. Good to see though that this seems to be a non-issue in the newer futures API.

Reply:
> with the current futures, do you see a better way for the general case of wanting to nest Sinks with different item types (due to being different layers of abstraction) while avoiding the re-encoding from the upper to the lower-level item type?

I don't think so. That's presumably why it has been changed in new futures.